### DISCUSSION TITLE
[select-rich] Implement @change event

### DISCUSSION THREAD

**Discussion by MathieuPuech** [ing-bank/lion#933] [General]
There is currently only `@model-value-changed` event on the select-rich component. I would like to have a `@change` event with the same behavior of native select change event.

**Comment by MathieuPuech** [comment]
I can create a PR if you agree with that

**Comment by MathieuPuech** [comment]
I think we may not use `change` for the event name since it's not a trusted event. What should be the name of this event?

**Comment by jorenbroekema** [comment]
Hmm I think in general you should use `@model-value-changed` instead of `@changed`. The reason why you can use `@changed` sometimes is probably because that custom input wraps a native input, which is not the case for a rich select. We could dispatch it ourselves, but I'm not sure what the added value would be over `@model-value-changed`, they would essentially be synonymous if we did this right?

**Reply by MathieuPuech** [reply]
There is some differences with the `@change` event, I made a quick demo of what could be the change event:

https://codesandbox.io/s/lion-select-rich-dz3ce?file=/src/my-select-rich.js

And a short video showing how change and model-value-changed events 
![2020-10-08_08-39-51](https://user-images.githubusercontent.com/9367152/95459613-ea198f80-0941-11eb-8b55-efe44b54d2b3.gif)

The change event is dispatched only when the interaction with the select is finished.

I've added the native `<select>` next to the `<lion-select-rich>` to see the differences

**Reply by jorenbroekema** [reply]
Aah so it's different in that it:
- model-value-changed fires initially (when it goes undefined --> 'red')
- model-value-changed fires every time when the checked index changes, whereas change only changes if you actually blur or if the dropdown was closed when checked index changed

I think we ought to mimic native behaviour as much as possible with model-value-changed, I don't think we want to send a second similar event. I'm on mac so I can't really check for windows right now, but the model-value-changed seems aligned on the 2nd point for mac. I will check tonight for windows.

@tlouisse  @daKmoR  do you guys know why we explicitly fire the model-value-changed initially even though the platform doesn't?

**Reply by MathieuPuech** [reply]
You're right, I've just tested on Mac, it's better for the model-value-changed. The only thing is there is 2 model-value-changed fired on load (mac safari). I don't know why there is 2 fired (that's another issue).

I didn't remember there was another issue related to the model-value-changed fired too much on windows:

https://github.com/ing-bank/lion/issues/708

**Reply by jorenbroekema** [reply]
Soo yeah I get the same as Thijs on 708, so if I understand correctly, we need to solve #708  and create another issue for initial model value changed fired twice on Safari, and then this discussion can be closed? Or is there something else?

We will soon have cross browser tests enabled in our CI, so then we should be able to create a failing test + issue for the initial model-value-changed fired twice on mac safari. Will let you know!

**Reply by daKmoR** [reply]
jup there is already a failing test for the model value if we run it on webkit 😅 

https://github.com/ing-bank/lion/pull/1021

**Reply by MathieuPuech** [reply]
Yes #708, another issue for safari but there is the initial event sent on all browsers that is maybe another issue too

**Reply by tlouisse** [reply]
@MathieuPuech @jorenbroekema @daKmoR Sorry for the very late response.
A simple way to solve #708  would be to suppress firing model-value-changed until the dropdown closes and fire an extra event on dropdown close.
But then we would have a situation where `mySelect.modelValue` would not be in sync.

Making modelValue not update for ChoiceGroups when opened would have quite a big impact on code, I think. And it would be breaking for people depending on this api (getting feedback about the fact that a user changed a value might be handy sometimes (on top, there is no way to 'cancel' the selection, which pledges for the case of getting this info as soon as possible)).
We would also have different behavior for a listbox and a dropdown listbox (or do we want to have the same behavior on blur of a listbox? And what about radio-groups/checkbox-groups?)

What we could also do is add meta info in the model-value-changed event, like we do for `isTriggeredByUser` as well: https://github.com/ing-bank/lion/pull/1167

So smth like: 
```html
<lion-select @model-value-changed="${myHandler}">
```
```js
function myHandler(ev) {
  if (!ev.detail.isChoiceInProgress) {
     // select changed
  }
}

```

There are probably better names for this :)

**Comment by jorenbroekema** [comment]
I think I'm moving this to discussion since it's not a bug but more a discussion or feature request
